airbyte-commons-protocol/src/main/java/io/airbyte/commons/protocol/VersionedProtocolSerializer.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.commons.protocol;
+
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.commons.version.Version;
+import io.airbyte.protocol.models.ConfiguredAirbyteCatalog;
+
+/**
+ * Serialize a ConfiguredAirbyteCatalog to the specified version
+ * <p>
+ * This Serializer expects a ConfiguredAirbyteCatalog from the Current version of the platform,
+ * converts it to the target protocol version before serializing it.
+ */
+public class VersionedProtocolSerializer implements ProtocolSerializer {
+
+  private final ConfiguredAirbyteCatalogMigrator configuredAirbyteCatalogMigrator;
+  private final Version protocolVersion;
+
+  public VersionedProtocolSerializer(final ConfiguredAirbyteCatalogMigrator configuredAirbyteCatalogMigrator, final Version protocolVersion) {
+    this.configuredAirbyteCatalogMigrator = configuredAirbyteCatalogMigrator;
+    this.protocolVersion = protocolVersion;
+  }
+
+  @Override
+  public String serialize(final ConfiguredAirbyteCatalog configuredAirbyteCatalog) {
+    return Jsons.serialize(configuredAirbyteCatalogMigrator.downgrade(configuredAirbyteCatalog, protocolVersion));
```

Comment:
Shy is `serialize` always calling `downgrade` and what happens if `protocolVersion` the latest version?

Reply:
The Migrator figure out the migration path, if the version is current, nothing happens.

Reply:
Is there a scenario under which `serialize` will have to upgrade rather than downgrade a catalog? E.g. if the connector supports only the newer version? And if that's the scenario - does `downgrade` then essentially upgrade instead of `downgrading`?